pkg/reconciler/pipeline/validate.go

```diff
@@ -89,6 +89,17 @@ func ValidatePipeline(pl *dfv1.Pipeline) error {
 		return fmt.Errorf("pipeline has no sink, at least one vertex with 'sink' defined is required")
 	}
 
+	for k, s := range sources {
+		if s.IsUDSource() {
+			if s.Source.UDSource.Container == nil || s.Source.UDSource.Container.Image == "" {
+				return fmt.Errorf("invalid user-defined source vertex %q, a customized image is required", k)
+			}
+			if s.Source.HTTP != nil || s.Source.Kafka != nil || s.Source.Nats != nil || s.Source.RedisStreams != nil || s.Source.Generator != nil {
```

Comment:
can you add a TODO to move `redis`, `nats`, to user-defined-source, and not in the core?

Reply:
Could you elaborate more on why?

Reply:
i want to move those away from our core numaflow code; our core code is getting bloated. i would like to do that with UDSource release (0.11).